Package setup

jmpressman opened the pull request:
Sets up smart select package

jmpressman committed:
Package setup

jmpressman commented:
@Tranquility I am tempted to separate the Single and Multi selects into separate modules. Primarily because the shared code between the two different types of selection requires a union type that in my opinion adds clutter and makes the end implementation a bit more complicated than it needs to be.

For now both selects are part of the same module, but I want to see if you think that union type is as clunky and unappealing as it feels to me.

Tranquility commented on README.md:
```diff
@@ -1,2 +1,138 @@
+    { internalMsg : ( Msg a, Maybe a ) -> msg
+    , optionLabel : a -> String
+    , optionDescription : a -> String
+    , searchFn : SearchUnion a
```
We can introduce LocalSettings and RemoteSettings so we don't need the union.

jmpressman replied:
Are you suggesting to have two separate search functions (each with their own settings)? If so, how would the select know which one to use?

I suppose both could be a Maybe function, but then we would have to handle the cases when neither and both functions are defined. Not the end of the world, but a little clunky in my opinion.

Like yourself, I've never been a fan of the search union, but I couldn't find a better alternative. Happy to discuss further next week.

Tranquility commented on README.md:
````diff
@@ -1,2 +1,138 @@
+4. Configure the settings for the smart select. The `Settings` type as defined in the `SmartSelectSingle` module is below:
+
+```elm
+type alias Settings msg a =
````
We can provide a function that returns settings with reasonable defaults set. Because I like this it looks very complicated. So this step should use the use the the default settings and refer to the docs for this type alias.

jmpressman replied:
I'm not sure I agree, apart from debounceDuration, characterSearchThreshold, and closeOnSelect, which can reasonably have defaults, all of the remaining settings fields are dependent on the implementer to define.

For example, internalMsg has to come from the implementer... That is what allows us to bypass the Html.mapping. If we want to ditch that convention entirely, then we don't need to have the internalMsg at all, but otherwise, it has to be defined outside of the select component.

Furthermore, optionLabel and optionDescription also need to be defined by the implementer because the select is not aware of the structure of the data it is displaying so it would be impossible to define a function in the select that could pull the relevant field from the data.

Finally, search is so specific to an implementer's use case for both local and external searches, we couldn't realistically define a default.

So... we could provide a function that expects the problem fields I've mentioned above and returns settings with default debounce, characterThreshold, and closeOnSelect values... but does that really add much? I like that the settings are super transparent at the moment, even if they are a bit verbose.

Tranquility commented on README.md:
```diff
@@ -1,2 +1,138 @@
+
+`SingleSelect.view` expects the following (in order):
+- a boolean indicating if the select is disabled or not
+- a string indicating the name of the data being selected i.e. "Product"
```
Why is this not in the settings?

jmpressman replied:
Good question. I can change this.

Tranquility commented on README.md:
```diff
@@ -1,2 +1,138 @@
+`SingleSelect.view` expects the following (in order):
+- a boolean indicating if the select is disabled or not
+- a string indicating the name of the data being selected i.e. "Product"
+- a string indicating the currently selected entity if any i.e "Vase"
```
What is this can this maybe be moved into the settings too?

jmpressman replied:
I should be able to move this into settings as well.

Tranquility commented on examples/index.html:
```diff
@@ -0,0 +1,38 @@
+    </div>
+    <script src="./Compiled/example.js"></script>
+    <script>
+      Elm.SmartSelectExample.Main.init({node: document.getElementById("example")});
```
I think it makes sense to split this into two apps (one for each example) to make each example easier to understand and make it easier to see what's really needed to get this running.

jmpressman replied:
Agreed. Will do.

mschindlerMM commented:
I just tested it out of pure interest and I really like it. But I have to say that I found it somewhat irritating to provide the Settings and the Data 3 times (once in the update, once in the view and once in the subscription) - especially because I had to handle my RemoteData every time as well. Why don't we just save the Settings and Data in the SmartSelect once set (e.g. via the init call)?

Great work!

jmpressman commented:
@mschindlerMM that's an interesting idea on the settings front! I will take a look at this today. As for the data, are you referring to the selection state? That might be a little more difficult to manage as the current architecture of the select is intentionally designed not to maintain the selection state to avoid "2 sources of truth".

I suppose a work around would be the following: instead of having the caller have a separate selected field in his or her model in addition to the smart select. The smart select *could* be the selected thing as well. So instead of:

```elm
type alias Model =
  { ...
  , selected : SomeDataType
  , smartSelect : SmartSelect.SmartSelect
  }
```
We would have:

```elm
type alias Model =
  { ...
  , selected : SmartSelect.SmartSelect -- selected now lives in here... use SmartSelect.getSelected to retrieve it.
  }
```

How does that sound?

jmpressman committed:
Update select settings type & separate examples

- add "optionType" to select settings
- single and multi select now have own examples
